### ISSUE TITLE
BuddyPress support

### ISSUE THREAD

**Issue by gnotaras** [gnotaras/wordpress-add-meta-tags#11]
I'm opening this issue for open discussion about the planned BuddyPress support in Add-Meta-Tags.

Please, feel free to post your ideas and feedback.

**Comment by gnotaras** [comment]
The current plan for AMT v2.9.7 is to have internal support for the generation of metadata for member pages.

If the *Extended Profiles* component has not been enabled, metadata based on the WordPress user profile will be generated. In fact, this is possibly the only case AMT will be able to generate metadata automatically for user profile pages.

However, in case the *Extended Profiles* component has been activated, it will be possible to filter and manually insert metadata for the fields you have added to the member profiles.

**Comment by schwarzaufweiss** [comment]
"…plans to let the admin choose whether Add-Meta-Tags should consider the BuddyPress member pages as the profile pages of the authors of posts…" – good plans. On my site I'm using this plugin:
https://wordpress.org/plugins/bp-blog-author-link/

**Comment by gnotaras** [comment]
Hello Hans, thanks for contributing to this discussion. This plugin will be a great help.

**Comment by gnotaras** [comment]
Finally, I think an extra *template tag* will be required as the settings page of v2.9.7 will have an extra selection box, where the user will be able to select the source of the local author profile. This option basically affects the URL that is used wherever a URL to a local author profile is required. The functionality is affected as described in [this help section](https://github.com/gnotaras/wordpress-add-meta-tags/blob/master/amt-admin-panel.php#L224). The template tag should be following this selection.

BTW, I checked the BuddyPress API documentation, but was disappointed to find out there is no such thing yet.. The source code is the only source of information. So, as a starting point I've implemented the [basic mechanism](https://github.com/gnotaras/wordpress-add-meta-tags/blob/master/metadata/amt_extended.php#L1270) that supports the insertion of BP metadata, but, adding any specific meta tags for the various components (even extended user profiles) will take some more time. In the meantime contributions about this are very welcome.

I've started writing a page on the wiki about [generating metadata for custom content](http://www.codetrax.org/projects/wp-add-meta-tags/wiki/Custom_Content). There will be a section which will describe how to insert/extend BuddyPress and bbPress metadata. I hope this helps.

**Comment by schwarzaufweiss** [comment]
I've just sent an email to Boone (BuddyPress core dev) with a link to this ticket. Hopefully he will find a minute to give you some hints...

**Comment by gnotaras** [comment]
Thanks, but I really don't mean to waste anyone's time with this. :) Time permitting, I'll be adding more support for the various components in the following weeks/months (starting with member pages and extended profiles). I'll also keep this issue open for possible user contributions.

From what I see, if no BP component is enabled, then the BP member page contains information from some of the WP user profile's fields. I guess this could be supported by AMT out of the box.

On the other hand, metadata support for BP's _Extended Profiles_ component requires more thought about how it should be implemented. Maybe, AMT should contain the necessary code to generate metadata for fields that are commonly found in profiles, such as location, contact info, fields of expertise etc, and also have a map (associative array) which would map the field names AMT supports with the actual field names that are used in a BP installation. The user would just have to override that map and provide AMT with the right associations of the fields.

These are just thoughts. I'll need to get more familiar with BP before proceeding.

George

**Comment by schwarzaufweiss** [comment]
Nothing around BuddyPress is "waste of time", at least as long as Facebook doesn't make people happy… ;-) What I wrote earlier: if you would like to have access to my BP site pls send an email to keller [at] fusion.ch and I will open an account for you.

**Comment by boonebgorges** [comment]
Hi all -

@gnotaras There is some documentation on BP's internal APIs on https://codex.buddypress.org, but you are correct that it's incomplete. Sorry about that - it's a large and ongoing project.

If you have implementation questions, and if I could save you some time by pointing you in the right direction, don't hesitate to ping me on this ticket.

**Comment by gnotaras** [comment]
Hi @boonebgorges 

Thanks for your reply. I've also started using the support forum at buddypress.org and so far my questions have been replied. I'm still finding my way around.

 I have no problem checking the source code. It requires some more time, but this is absolutely fine. My only concern about the BP Codex being incomplete is whether this also means that the BP API should not be considered final yet.

Kind Regards,
George

**Comment by schwarzaufweiss** [comment]
Only a simple idea: at least for me it would be ok to see "Description" from /members (or whatever permalink that page has) in a search engine index for a dynamically generated member page (/members/username). At the moment it's something strange like "@username. active ... username and username are now friends, 17 hours ago …" etc. – same for /activity etc. Other words: would it be possible to "fall back" to the static parent page for dynamically generated BP pages? /members, /activity and /groups are predefined wp pages (/admin.php?page=bp-page-settings) where it's possible to place metadata…

**Comment by gnotaras** [comment]
If I understand correctly, wouldn't that lead to duplicate descriptions on the `/member/username/` pages? The plan was to use the user's bio (or at least a part of it) as the description that is used in the metadata.

BTW, currently I've noticed that AMT generates some totally unexpected metadata on some BP pages. At first I thought that it was AMT goofing up, but later I realized it really depends on the queried object on each page. There are still some things I haven't figured out about BP yet. The bottom line is that AMT must support BP before the metadata starts making sense. Otherwise any generated metadfata must be filtered out on BP pages.

BTW, I've gathered much of the needed information which will help me implement the metadata generators on member pages. I'll try to do some work tonight.

**Comment by schwarzaufweiss** [comment]
These "totally unexpected metadata" are the reason why I thought it would be better to show something more meaningful. Yes, if there is some user's bio (description in user profile), that would be perfect – but  if not, site_description is definitely better than "scrambled eggs".

**Comment by gnotaras** [comment]
> if there is some user's bio (description in user profile), that would be perfect – but if not, site_description is definitely better than "scrambled eggs".

This sounds like a good final fallback description if all else fails.

> That's a non-standard field. On a out-of-the-box BP installation that field will definitely not exist.

If the _Extended Profiles_ component is enabled, then apart from the member's full name and avatar, it is impossible for the plugin to guess the fields that are being used. However, I plan to provide a filterable array which contains `field_amt_understands => field_actually_used` mappings. By default AMT will be able to generate meta tags for some information that is commonly found in profiles. If the fields that are actually used are different than the fields AMT supports, then this could be solved by providing a proper field mapping.

I'd have done some work yesterday, but I realized that the custom content detection mechanism I've implemented has a small problem and needs to be modified. I'll keep you updated about this and I'll most probably make a beta release available once things seem to be working.

**Comment by schwarzaufweiss** [comment]
Cool. I should have thought about this. If you can show a selector in AMT settings the moment BP is active to choose the xprofile bio field – or whatever AMT should look for – that would be great.

**Comment by gnotaras** [comment]
Adding an extra option for specific BP features will need some more thought.. I'm generally hesitant to add options for so specific features, but we'll see if something like that would be required. On the contrary, making such customisations programmatically by adding some code in the `functions.php` file of the theme or in a [custom plugin](http://www.codetrax.org/projects/wp-add-meta-tags/wiki/Cookbook_Snippet_Uses#Custom-Plugin), which is theme agnostic, is the preferred way for maximum flexibility.

**Comment by gnotaras** [comment]
Hello Hans,

I'm currently working on the opengraph metadata of BP profiles. If extended profiles are enabled, the following fields will be queried for the user's description: `description, bio, about`

If one is not found, the, then the generic description `Profile of USER_FULL_NAME` will be used.

The field list that is queried is part of a field map AMT will be using in order to retrieve the user info, the generation of metadata of which it supports. This map can be overridden by the user to support alt field names.

I'll possibly push a pre-release tonight that will support the `description` meta tag (and possibly the `keywords` meta tag based on an `interests` field) and opengraph metadata. Not sure If I'll be able to work on schema.org today.

Your feedback is welcome.

**Comment by gnotaras** [comment]
Hi all,

The alpha release `2.9.7a1` is now available at: http://www.codetrax.org/projects/wp-add-meta-tags/files

In this release, basic, Opengraph and Twitter Cards metadata is generated for BP member profiles. If the _Extended Profiles_ component has not been enabled, user information from the WordPress user profile is used.

If the _Extended Profiles_ component is activated, then metadata is generated based on field mappings of the form:

```
field_amt_understands => array_of_possible_fields__which_are_actually_used
```

Actual function at the time of writing:

```
function amt_buddypress_get_xprofile_field_map() {
    $xprofile_field_map = array(
        'description'       => array('description', 'bio', 'about'),
        'keywords'          => array('keywords', 'skills', 'interests'),    // TODO: Future: add group names?
        'first_name'        => array('first name'),
        'last_name'         => array('last name'),
        'gender'            => array('gender', 'sex'),
    );
    return apply_filters( 'amt_buddypress_xprofile_field_map', $xprofile_field_map );
}
```

The above mappings can be overridden by attaching a filtering function to the `amt_buddypress_xprofile_field_map` filter hook.

The Basic metadata generator generates a `description` meta tag and a `keyword` meta tag (only via a field mapping in extended profiles)

The Opengraph metadata is almost complete. Meta tags that are missing:

 * `fb:profile_id` (still needs some thought)
 * `og:see_also` (Perhaps add Facebook, Twitter, Google+ profile URLs from AMT fields of the WordPress user profile as 'og:see_also' meta tags)
 * `profile:gender` (Will be supported in extended profile only via field mapping)
 * `profile:username` (No plans for it)

The Twitter Cards generator generates a basic card for now.

Currently no Dublin Core and no Schema.org metadata.

Instructions how to use:

1. Enable the BuddyPress support in the plugin settings.
1. (optional) Set the new option about the source of local author profiles to `BuddyPress`.
1. (optional) Use the `amt_local_author_profile_url()` template tag to generate a URL to the post author's profile based on the selection of the source of local profiles.

Please report any issues.

The generators for the missing metadata will be implemented at a slow pace during the upcoming weeks and new pre-release versions will be announced here.

Big thanks to all for the help so far.

George

**Comment by gnotaras** [comment]
Last but not least, please feel free to [customize the generated BuddyPress metadata](http://www.codetrax.org/projects/wp-add-meta-tags/wiki/Custom_Content#Metadata-for-BuddyPress) and post your customizations. Thanks.

**Comment by gnotaras** [comment]
I've released `2.9.7a2` which fixes several issues. Opengraph support is now complete (with the exception of `fb:profile_id`, which still needs more thought).

**Comment by gnotaras** [comment]
Noticed that there is a problem with the display of a BP user's profile, if the Schema.org microdata generator is enabled. More investigation is required. Please use these alpha releases in testing environments.

**Comment by gnotaras** [comment]
I've released `2.9.7a3` which fixes the aforementioned issue with the display of the BP user profile if Schema.org microdata is enabled.

This `a3` release also adds basic support for Schema.org microdata & json+ld for BP user profiles.

The schema.org generators will support more fields in future releases and also the visibility of the fields will be taken into consideration.

Your feedback is welcome.

**Comment by gnotaras** [comment]
The first beta release `2.9.7b1` is out.

**Comment by gnotaras** [comment]
The stable `2.9.7` release is out.

Currently, the plugin does not support the customization of the content of the `title` HTML element of BP profile pages, so the title generated by BP (in fact this is the blog's name!) is still used as is. Future releases will add title customization support and also extend the Schema.org generators.

Feel free to post your feedback and customizations.

George

**Comment by gnotaras** [comment]
Beta version `2.9.8b1` containing various fixes and additions has been released.

Some of the changes:

 * The public profile URL is now generated taking into account BP's profile slug.
 * The privacy settings are now checked on a per field basis.
 * The schema.org generators have been updated. More fields are supported. (Please check the 'field map' in the linked wiki page below for more info)

The documentation on the wiki has been updated:

http://www.codetrax.org/projects/wp-add-meta-tags/wiki/Custom_Content#Metadata-for-BuddyPress

Your feedback is welcome.

George

**Comment by schwarzaufweiss** [comment]
we got some fatal errors…
…you have access to the log in Dashboard of lerngruppen.net

--
HansRuedi Keller, Haldenstr. 28, 8904 Aesch ZH, Switzerland, +41 79 716 11 11

**Comment by gnotaras** [comment]
Hans, I've just released `2.9.8b2` which fixes this issue. I was not aware `bp_get_profile_slug()` was such a recent addition.

George

**Comment by gnotaras** [comment]
I'll now remove the `needs-feedback` tag from this issue, since no additional metadata support is planned for the time being (v2.9.10 at the time of writing).

However, improvements of the profile metadata will take place as needed. Your feedback is welcome.

George
